### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by madsmty**
Carpetas para darle orden al codigo
Elaboracion de routers
Llamada a Axios
Generacion de JWT
Verificacion de JWT
creacion de modelo de usuarios

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -1,60 +1,24 @@
 import express, {Application, NextFunction, Request, Response} from "express";
-import axios, {AxiosResponse} from "axios";
+import { logToDB } from "./middleware/logToDB";
+
 const app: Application = express();
 const port: number = 3000;
 
+const usersRoutes = require('./routes/users');
```
Comment (danielavaladez) [inline src/app.ts]: `require` isn't used in Typescript. Use `import` instead.

File: src/controllers/login.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import express, { Router,NextFunction, Request, Response } from 'express';
+const jwt = require('jsonwebtoken');
+
+exports.get_jwt = (req: Request, res: Response, next: NextFunction)=> {
```
Comment (danielavaladez) [inline src/controllers/login.ts]: This should be a typescript class. 

```
export class Login {
   async get_jwt() {

}
}
```

File: src/controllers/users.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import express, { Router,NextFunction, Request, Response } from 'express';
+import {callExtAPI} from '../services/callExtAPI';
+import { AxiosResponse } from 'axios';
+
+exports.users_get_all = async (req: Request, res: Response, next: NextFunction)=>  {
```
Comment (danielavaladez) [inline src/controllers/users.ts]: Change to Class as well.

File: src/middleware/checkJWT.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import express, { Router,NextFunction, Request, Response } from 'express';
+const jwt = require('jsonwebtoken');
+
+module.exports = (req: Request ,res:Response,next: NextFunction) => {
```
Comment (danielavaladez) [inline src/middleware/checkJWT.ts]: Let's re-declare this as:

```
export const jwtValidator (... )= {
...
}
```

File: src/middleware/logToDB.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+import express, { Router,NextFunction, Request, Response } from 'express';
+
+export function logToDB(req: Request, res: Response, next: NextFunction) {
```
Comment (danielavaladez) [inline src/middleware/logToDB.ts]: Change to class.

File: src/services/callExtAPI.ts [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import axios, {AxiosResponse} from "axios";
+
+export async function callExtAPI<T>(url:string): Promise<AxiosResponse> {
```
Comment (danielavaladez) [inline src/services/callExtAPI.ts]: Change to class.

**Review by danielavaladez — CHANGES_REQUESTED** [review]

File: node_modules/.package-lock.json
Comment (danielavaladez) [inline node_modules/.package-lock.json:1]: this should not be part of the PR

File: build/app.js
Comment (danielavaladez) [inline build/app.js:1]: this should not be part of the PR

File: src/middleware/logToDB.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+import { NextFunction, Request, Response } from 'express';
+import { LogEntry } from '../models/logEntry';
+import jwt from 'jsonwebtoken';
+import { TokenInterface } from '../classes/token';
+
+
+export async function logToDB(req: Request, res: Response, next: NextFunction) {
+    const now= new Date();
```
Comment (danielavaladez) [inline src/middleware/logToDB.ts]: Let's keep the handler simple and move all the implementation details to a controller typescript class.

File: src/controllers/users.ts [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+
+//axios.interceptors.response.use((r: ServerResponse) => r.data);
+
+export const users_get_all = async (req: Request, res: Response, next: NextFunction)=>  {
+    const url:string = "https://jsonplaceholder.typicode.com/users/";
+    try {
+        const resultArray:ServerResponse = await axios.request({
+            url:url,
```
Comment (danielavaladez) [inline src/controllers/users.ts]: `url: url,` => `url,`

File: src/controllers/users.ts
Comment (danielavaladez) [inline src/controllers/users.ts:1]: This file would need to either be changed to a `UserService` class with implementation methods for each of the API calls.

**Review by danielavaladez — CHANGES_REQUESTED** [review]

File: src/app.ts (hunk 1 above)
Comment (madsmty) [reply src/app.ts]: Done

File: node_modules/.package-lock.json
Comment (madsmty) [reply node_modules/.package-lock.json:1]: Done

File: build/app.js
Comment (madsmty) [reply build/app.js:1]: Done

File: src/controllers/users.ts (hunk 8 above)
Comment (madsmty) [reply src/controllers/users.ts]: Done

File: src/middleware/checkJWT.ts (hunk 4 above)
Comment (madsmty) [reply src/middleware/checkJWT.ts]: done

File: src/controllers/users.ts (hunk 3 above)
Comment (madsmty) [reply src/controllers/users.ts]: Done

File: src/services/callExtAPI.ts (hunk 6 above)
Comment (madsmty) [reply src/services/callExtAPI.ts]: Done (functionality moved to Controllers)

File: src/controllers/users.ts
Comment (madsmty) [reply src/controllers/users.ts:1]: Done

File: src/controllers/login.ts (hunk 2 above)
Comment (madsmty) [reply src/controllers/login.ts]: Done

File: src/middleware/logToDB.ts (hunk 7 above)
Comment (madsmty) [reply src/middleware/logToDB.ts]: done

File: src/middleware/logToDB.ts (hunk 5 above)
Comment (madsmty) [reply src/middleware/logToDB.ts]: done

File: src/controllers/posts.ts [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    };
+    //userMap.id = element.id
+    return postMap;
+});
+return responseArray
+}
+
+    async getUserPosts():Promise<boolean> {    
+        let userId:string = this.req.params.userId;
+        const userUrl:string = this.domain + "users/" + userId;
```
Comment (danielavaladez) [inline src/controllers/posts.ts]:
```
const userUrl: string = `${this.domain}users/${userId}`
```
User string interpolation instead of `+`

File: src/controllers/posts.ts [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    req:Request;
+
+    constructor (res:Response, req: Request) {
+        this.res = res;
+        this.req = req;
+    }
+
+convertPosts(paramUser: ServerUserElement,paramPostArray: ServerPostArray):Array<any> {
+    //console.log(paramUserArray.address)
+    let responseArray = paramPostArray.data.map((element)=>{
```
Comment (danielavaladez) [inline src/controllers/posts.ts]: User `lint` to find out what `let`s need to be turned into `const`.

File: src/controllers/login.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    req:Request;
+
+    constructor (res:Response, req: Request) {
+        this.res = res;
+        this.req = req;
+    }
+
+    loginUser():boolean {
+        // JWT with regular user role
+         const token = jwt.sign( {
```
Comment (danielavaladez) [inline src/controllers/login.ts]: Use `prettier` to format the code properly.

File: src/controllers/posts.ts [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+        let userId:string = this.req.params.userId;
+        const userUrl:string = this.domain + "users/" + userId;
+        //const userUrl:string = this.domain + "users?id=" + userId;
+        const postUrl:string = this.domain + "users/" + userId + "/posts";
+        let functionStatus:boolean = true;
+        let responseArray:Array<any>;
+        //responseArray.data = [];
+        try {
+            console.log("- API Call")
+            const resultUser:ServerUserElement = await axios.request({
```
Comment (danielavaladez) [inline src/controllers/posts.ts]: `const [resultUser, resultPostArray] = await Promise.all([promise1, promise2]);`

File: src/controllers/posts.ts [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+            //console.log(responseArray)
+            console.log("Post Response Sent");      
+            this.res.status(200).json(
+                responseArray
+            ); 
+        }
+        catch(error) {
+            console.log(error);
+            functionStatus = false;
+            throw new Error(`Error in 'getUserPosts': ${error}`);
```
Comment (danielavaladez) [inline src/controllers/posts.ts]: If you throw here, the handler will error out because there is no error handling in the handler.

**Review by danielavaladez — CHANGES_REQUESTED** [review]
Let's remove all of the `//comments` as well.

File: src/controllers/posts.ts (hunk 9 above)
Comment (madsmty) [reply src/controllers/posts.ts]: done

File: src/controllers/posts.ts (hunk 10 above)
Comment (madsmty) [reply src/controllers/posts.ts]: done

File: src/controllers/login.ts (hunk 11 above)
Comment (madsmty) [reply src/controllers/login.ts]: done

**Comment by danielavaladez** [comment]
Please raise another PR for the two remaining unresolved comments.
